builtin/game/anticheat.lua

```diff
@@ -0,0 +1,288 @@
+	end
+	def.enable = function(self)
+		if self.on_enable then
+			if self:on_enable() ~= false then
+				self.enabled = true
+			end
+		else
+			self.enabled = true
+		end
+	end
+
+	if def.disable then
+		core.log("warning", "Anticheat object "..name.." definition contains a disable method. Use on_disable instead. Overwriting.")
+	end
+	def.disable = function(self)
+		if self.on_disable then
+			if self:on_disable() ~= false then
+				self.enabled = false
+			end
+		else
+			self.enabled = false
+		end
+	end
+
+	core.registered_anticheat_objects[name] = def
+end
+
+function core.get_anticheat_object(player, type)
+	if type(player) == "string" then
+		player = minetest.get_player_by_name(player)
+	end
+
+	local player_name = player:get_player_name()
+	if not current_anticheat_objects[player_name] then
+		current_anticheat_objects[player_name] = {}
+	end
+	if not current_anticheat_objects[player_name][type] then
+		if not core.registered_anticheat_objects[type] then
+			return nil
+		end
+
+		-- Create anticheat object
+		local obj = {}
+		setmetatable(obj, {__index = core.registered_anticheat_objects[type]})
+		obj.player = player
+		--obj.enabled = core.settings:get_bool("enable_anticheat") and not core.is_singleplayer()
+		-- DEBUG:
+		obj.enabled = core.settings:get_bool("enable_anticheat")
+		if obj.init then
+			obj:init()
+		end
+		current_anticheat_objects[player_name][type] = obj
+	end
+
+	return current_anticheat_objects[player_name][type]
+end
+
+function core.run_anticheat_callbacks(player, type, params)
+	assert(core.registered_anticheat_objects[type])
+	if not params then
+		params = {}
+	end
+	params.type = type
+	return core.run_callbacks(core.registered_on_cheats, 0, player, params)
+end
+
+core.register_on_leaveplayer(function(player)
+	local player_name = player:get_player_name()
+	if not current_anticheat_objects[player_name] then
+		return
+	end
+	for type, obj in pairs(current_anticheat_objects[player_name]) do
+		if obj.deinit then
+			obj:deinit()
+		end
+	end
+	current_anticheat_objects[player_name] = nil
+end)
+
+-- Builtin anticheat definitions. These anticheat objects are called from the C++ engine
+
+core.register_anticheat_object("interacted_too_far", {
+	check = function(self, node_pos)
+		if not self.enabled then
+			return false
+		end
+
+		local distance = vector.distance(self.player:get_pos(), node_pos)
+		local max_distance = self:get_max_distance()
+
+		if distance > max_distance then
+			core.log("action", "Player " .. self.player:get_player_name()
+						.. " tried to access " .. core.pos_to_string(node_pos)
+						.. " from too far: "
+						.. "d=" .. distance .. ", max_d=" .. max_distance
+						.. ". ignoring.")
+			core.run_anticheat_callbacks(self.player, "interacted_too_far", {node_pos = node_pos, distance = distance, max_distance = max_distance})
+			return true
+		else
+			return false
+		end
+	end,
+
+	set_max_distance = function(self, max_distance)
```

Comment:
I can't think of any reason a mod might want to set a higher range for server-side <b>only</b> (the client won't even know how to use this).